Subspaces types

bragaz opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description
This PR introduce the `x/subspaces` types.

## Checklist
- [x] Targeted PR against correct branch.
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Wrote unit tests.
- [ ] Wrote integration tests (simulation & CLI). 
- [ ] Updated the documentation. 
- [ ] Added an entry to the `CHANGELOG.md` file.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.

RiccardoM reviewed (commented):
Can you also add comments to all Proto fields?

bragaz commented on proto/desmos/subspaces/v1beta1/msgs.proto:
```diff
@@ -0,0 +1,206 @@
+// ___________________________________________________________________________________________________________________
+
+// MsgEditSubspace represents the message used to edit a subspace fields
+message MsgEditSubspace {
```
I was thinking the same and I stopped while thinking about how to handle a default value of it, if the user doesn't specify it while making the tx.
Suppose we have an open subspace and we want to edit only its owner.   
What if the default value of the `--open` flag is `false`?
It will edit the open field anyway even though the user doesn't want that.
Do you think we can use a flag type `string` and converting it later inside the message itself to solve this?

RiccardoM replied:
@bragaz You're right, I didn't think about this. What I think we can do to solve this is, instead of using an `open` field, we can use a `type` field. We can then have `SUBSPACE_TYPE_OPEN` and `SUBSPACE_TYPE_CLOSED` to identify the various types of subspaces. These should be implemented as an `enum` similar to [the `x/gov` one inside the Cosmos SDK](https://github.com/cosmos/cosmos-sdk/blob/master/proto/cosmos/gov/v1beta1/gov.proto#L17)

RiccardoM commented on proto/desmos/subspaces/v1beta1/subspace.proto:
```diff
@@ -0,0 +1,74 @@
+  option (gogoproto.goproto_enum_prefix) = false;
+
+  // SUBSPACE_TYPE_UNSPECIFIED identifies an unspecified type of subspace (used in errors)
+    SUBSPACE_TYPE_UNSPECIFIED = 0 [(gogoproto.enumvalue_customname) = "Unspecified"];
```
```suggestion
  SUBSPACE_TYPE_UNSPECIFIED = 0 [(gogoproto.enumvalue_customname) = "Unspecified"];
```

RiccardoM commented on x/staging/posts/keeper/keeper_bench_test.go:
```diff
@@ -11,7 +11,7 @@ import (
 )
 
-// RandomPost returns a post with a 50% chance to have random medias and random poll
+// RandomSubspace returns a post with a 50% chance to have random medias and random poll
```
```suggestion
// RandomPost returns a post with a 50% chance to have random medias and random poll
```

RiccardoM commented on x/staging/subspaces/types/errors.go:
```diff
@@ -0,0 +1,17 @@
+const RootCodespace = "subspaces"
+
+var (
+	// ErrInvalidSubspace is returned if a subspace id is not valid
+	ErrInvalidSubspace = sdkerrors.Register(RootCodespace, 1, "invalid subspace")
```
Probably we can rename this to something like `ErrInvalidSubspaceID`?

bragaz replied:
yeah it will be better

RiccardoM commented on x/staging/subspaces/types/errors.go:
```diff
@@ -0,0 +1,17 @@
+import sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+// RootCodespace is the codespace for all errors defined in this package
+const RootCodespace = "subspaces"
```
I think we can remove this and use `ModuleName` instead

RiccardoM reviewed (commented):
I'm having a really hard time deciding whether or not we should really have the `NameParams`. Subspaces will be limited and ideally the name will be never used elsewhere except when the user might want to know it. Should we really cap it?

bragaz commented:
> I'm having a really hard time deciding whether or not we should really have the NameParams. Subspaces will be limited and ideally the name will be never used elsewhere except when the user might want to know it. Should we really cap it?

I added them when I was adding a check on the `name` length inside the `MsgCreateSubspace` and `MsgEditSubspace` and remembered we did a similar thing with `profiles` parameters, first we had the hard coded checks, then we switched to `params`. I know that the usage will be different, so yes, we can replace them with some checks inside each of the `ValidateBasic()` methods. Let me know what d you think @RiccardoM .

RiccardoM commented:
@bragaz I think we can simply remove all checks from subspace names. We should not have a min/max length as well as any kind of check of the values themselves (eg. regex). Allowing people to create subspaces with arbitrary names won't be a problem as far as I can think of.

RiccardoM reviewed (approved):
Changes look great. Can you please create a new branch (maybe something like `leonardo/manageable-subspaces` and use that as the target branch of this PR instead of `master`? You can change it without having to close the PR

bragaz commented:
@RiccardoM done
